compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEFusePackIntoForLoop.cpp

```diff
@@ -0,0 +1,101 @@
+// Copyright 2024 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree-amd-aie/Transforms/Passes.h"
+#include "mlir/Dialect/SCF/Transforms/TileUsingInterface.h"
+#include "mlir/IR/Iterators.h"
+#include "mlir/Pass/Pass.h"
+
+#define DEBUG_TYPE "iree-amdaie-fuse-pack-into-for"
+
+namespace mlir::iree_compiler::AMDAIE {
+
+namespace {
+
+class AMDAIEFusePackIntoForLoopPass
+    : public impl::AMDAIEFusePackIntoForLoopBase<
+          AMDAIEFusePackIntoForLoopPass> {
+ public:
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<scf::SCFDialect>();
+  }
+
+  AMDAIEFusePackIntoForLoopPass() = default;
+  AMDAIEFusePackIntoForLoopPass(const AMDAIEFusePackIntoForLoopPass &pass){};
+  void runOnOperation() override;
+};
+
+void AMDAIEFusePackIntoForLoopPass::runOnOperation() {
+  MLIRContext *context = &getContext();
+  func::FuncOp funcOp = getOperation();
+  IRRewriter rewriter(context);
+
+  // Walk through the graph in post order and find the pack op.
+  TilingInterface packOp;
+  funcOp->walk<WalkOrder::PostOrder, ReverseIterator>([&](TilingInterface op) {
+    if (isa<linalg::FillOp>(op)) {
+      packOp = op;
+      return WalkResult::interrupt();
+    }
+    return WalkResult::advance();
+  });
+  if (!packOp) {
+    LLVM_DEBUG(llvm::dbgs() << "There is no pack op to be fused.\n");
+    return;
+  }
+
+  // Search the first use by a scf::ForOp user.
+  scf::ForOp forOp;
```

Comment:
THis is kind of backwards. If you want to fuse the producers of the pack inside the loop, I am assuming you are only looking at lhs and rhs operands of the compute operation. 

What you should do instead is look for the compute op, if its operands are slices and the source of the slices are pack, fuse them in... 

Though, i have been thinking about this. A cleaner way to do this would be to just do the packing within the scf.for op and then just hoist the pack for the output operands out... conversely you should "hoist" the unpack operation out as well (but it goes after the loop). I havent fully worked this out though. I need some time to play with the transform dialect script you shared :P

Reply:
Thanks for taking a look into this. I'll push up another PR that using this pass to get to what AIR and AIE hardware requires. We can then evaluate the benefit of this pass. At the same time, I modified the logic a little bit to match what you described.